pkg/distributor/otlp/cmd/generate/main.go

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+
+	// Dummy dependency to ensure these files get vendored and we can use them as transformation sources.
+	_ "github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite"
+)
+
+func main() {
+	app := kingpin.New("generate", "Generate OTLP translation code.")
+
+	generate := func(*kingpin.ParseContext) error {
+		sources, err := copySources()
+		if err != nil {
+			return err
+		}
+		if err := patchSources(sources); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	app = app.Action(generate)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+}
+
+func patchSources(sources []string) error {
+	if out, err := exec.Command("gopatch", "-p", "cmd/generate/mimirpb.patch", "./").CombinedOutput(); err != nil {
+		return fmt.Errorf("failed to execute gopatch: %s", out)
+	}
+	sed := "sed"
+	if runtime.GOOS == "darwin" {
+		sed = "gsed"
+	}
+	for _, fname := range sources {
+		if out, err := exec.Command(sed, "-i", "s/PrometheusConverter/MimirConverter/g", fname).CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute sed: %s", out)
+		}
+		if out, err := exec.Command(sed, "-i", "s/Prometheus remote write format/Mimir remote write format/g", fname).CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute sed: %s", out)
+		}
+		if out, err := exec.Command("goimports", "-w", "-local", "github.com/grafana/mimir", fname).CombinedOutput(); err != nil {
+			return fmt.Errorf("failed to execute goimports: %s", out)
+		}
+	}
+
+	return nil
+}
+
+func copySources() ([]string, error) {
+	dpath := filepath.Join("..", "..", "..", "vendor", "github.com", "prometheus", "prometheus", "storage", "remote", "otlptranslator", "prometheusremotewrite")
+	entries, err := os.ReadDir(dpath)
+	if err != nil {
+		return nil, fmt.Errorf("list files in %q: %w", dpath, err)
+	}
+
+	var sources []string
+	for _, e := range entries {
+		if e.IsDir() || !strings.HasSuffix(e.Name(), ".go") || e.Name() == "timeseries.go" || strings.HasSuffix(e.Name(), "_test.go") {
+			continue
+		}
+
+		dst := fmt.Sprintf("%s_generated.go", e.Name()[:len(e.Name())-3])
+		if err := copyFile(filepath.Join(dpath, e.Name()), dst); err != nil {
+			return nil, err
+		}
+
+		sources = append(sources, dst)
+	}
```

Comment:
> There doesn't seem to be anything advanced here, just lot of error-checking boilerplate code

@pstibrany this is the complicated part, I'm not sure how I would rewrite it as shell script. Also, what do we gain from the rewriting effort? Are there any objective gains? I think it just becomes less understandable/maintainable.

Reply:
>  Also, what do we gain from the rewriting effort?

We will need to maintain 20-lines script instead of 200 lines Go program.

Reply:
I personally prefer maintaining the equivalent Go - how would you rewrite this in shell script though? I'm not even sure how, I'm not a shell expert.

Reply:
This is full script that does exact same job. I've verified it by removing _generated.go files, calling the script and checking if `git diff` shows anything.

```bash
#!/bin/bash

# Use GNU sed on MacOS falling back to `sed` everywhere else
SED=$(which gsed || which sed)

FILES=$(find ../../../vendor/github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite -name '*.go' ! -name timeseries.go ! -name "*_test.go")

for SRC in $FILES
do
  BASENAME=$(basename "$SRC")
  DST="${BASENAME%%.go}_generated.go"

  rm -f "$DST"
  echo "Processing $SRC to $DST"
  printf "// Code generated from Prometheus sources - DO NOT EDIT.\n\n" >"$DST"
  cat "$SRC" >> "$DST"

  gopatch -p cmd/generate/mimirpb.patch "$DST"

  $SED -i "s/PrometheusConverter/MimirConverter/g" "$DST"
  $SED -i "s/Prometheus remote write format/Mimir remote write format/g" "$DST"
  goimports -w -local github.com/grafana/mimir "$DST"
done
```

Console output:
```console
% pwd
/Users/peter/Grafana/mimir/pkg/distributor/otlp
% rm *_generated.go
% ./generate.sh
Processing ../../../vendor/github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite/histograms.go to histograms_generated.go
Processing ../../../vendor/github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite/number_data_points.go to number_data_points_generated.go
Processing ../../../vendor/github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite/helper.go to helper_generated.go
Processing ../../../vendor/github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go to metrics_to_prw_generated.go
Processing ../../../vendor/github.com/prometheus/prometheus/storage/remote/otlptranslator/prometheusremotewrite/otlp_to_openmetrics_metadata.go to otlp_to_openmetrics_metadata_generated.go
% git diff
```

Reply:
I replaced the current generator with your shell script, in pkg/distributor/otlp (putting it under tools/ is too much of a hassle IMO, see comment above).

Reply:
Thank you Arve!